pkg/manager/internal.go

```diff
@@ -135,22 +144,47 @@ func (cm *controllerManager) Add(r Runnable) error {
 		return err
 	}
 
-	var shouldStart bool
+	if cm.leaderElectionRunnables == nil {
+		cm.leaderElectionRunnables = make(map[string][]Runnable)
+	}
 
 	// Add the runnable to the leader election or the non-leaderelection list
-	if leRunnable, ok := r.(LeaderElectionRunnable); ok && !leRunnable.NeedLeaderElection() {
-		shouldStart = cm.started
-		cm.nonLeaderElectionRunnables = append(cm.nonLeaderElectionRunnables, r)
+	if leRunnable, ok := r.(LeaderElectionRunnable); ok && leRunnable.NeedLeaderElection() {
+		runnables := []Runnable{r}
+		leID := leRunnable.GetID()
+
+		// Check that leader election ID is defined
+		if leID == "" {
+			return errors.New("LeaderElectionID must be configured")
+		}
+
+		rs, exists := cm.leaderElectionRunnables[leID]
+		if exists {
+			runnables = append(runnables, rs...)
+		}
+
+		cm.leaderElectionRunnables[leID] = runnables
+
+		if cm.started {
+			// If Leader Election ID is already used and elected, start the controller.
+			// If it's appeared first time, start leader election for it.
+			if elected, _ := cm.electedLeaderElectionIDs[leID]; elected {
+				go func() {
+					cm.errChan <- r.Start(cm.internalStop)
```

Comment:
can we ever have a case where this races with losing the lease?

Reply:
Currently, if lease is lost we close le-group channel so runnable is started with closed channel. It's expected that runnable will exit immediately if stop channel is closed(this is the case with `controller.Controller` right now)